### PR TITLE
[SYCL][Graph] Expand on design doc

### PR DESCRIPTION
**Pull request by EwanC**
* Add sections to the design doc about the implementation of nodes & edges, as well as how the graph is linearized.

* convert our table of PI entry-points to UR and insert references to the rendered UR documentation.

* Information about our Level Zero adapter implementation added using mermaid diagrams.

* Consistent hyphenation of "command-buffer", "command-list", and "command-graph".

### CODE REVIEW

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:48]: Need to double check this link when https://github.com/oneapi-src/unified-runtime/pull/591/ is merged, won't be valid until then

Comment (EwanC) [reply sycl/doc/design/CommandGraph.md:48]: Live now and works 😃

File: sycl/doc/design/CommandGraph.md
Comment (mfrancepillois) [inline sycl/doc/design/CommandGraph.md]: A direct link to the file could be probably clearer that a link to the PR.

File: sycl/doc/design/CommandGraph.md
Comment (mfrancepillois) [inline sycl/doc/design/CommandGraph.md]: This sentence is difficult to understand. 
What CGF stand for?
Maybe some extra punctuation could help understanding. e.g. ... and after finalization, the CG ...

File: sycl/doc/design/CommandGraph.md
Comment (mfrancepillois) [inline sycl/doc/design/CommandGraph.md:134]: I'm not sure to understand this sentence.
In "Requirements" Section you mention that only Level Zero supports command-buffer, but here it seems that UR can enable other backends.

File: sycl/doc/design/CommandGraph.md
Comment (mfrancepillois) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
    id1[Barrier on EL that signals CB WaitEvent when completed]
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [reply sycl/doc/design/CommandGraph.md]: Makes sense, we can update this once the PR is merged.

File: sycl/doc/design/CommandGraph.md
Comment (julianmi) [inline sycl/doc/design/CommandGraph.md]: These sentences are not clear to me.

File: sycl/doc/design/CommandGraph.md
Comment (julianmi) [inline sycl/doc/design/CommandGraph.md]: Should we clarify what we are sorting? E.g. "Node Sorting" or "Execution Order".

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
`sycl::handler` from a queue submission is finalized, if the queue the
```

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]: I think it could be worth clarifying in this section that we're preventing the `CG` object from being submitted for execution as normal inside `handler::finalize()` and then storing it in the graph as a new node instead.

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]: Might be worth noting here that this is in keeping with the existing behaviour of the handler/normal queue submissions.

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
Buffers and accessors are supported in a command-graph. The following restrictions
```

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
- The lifetime of a buffer with host data will be extended by copying the underlying
```

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
- Host accessors on buffers that are currently used by a command-graph are prohibited.
```
I think the wording here was a bit ambiguous in that it could suggest that a buffer which had been used in a command-graph but that command-graph has since been destroyed would be prohibited, which I don't think we want.

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]: I don't know if this is too nitpicky but this might be misleading a reader to think that we actually create a new command list manually every time for these when we are getting them from the UR Queue which will likely just reuse one, and only create a new one in the worst case.

**Review by Bensuo — APPROVED** [review]
LGTM just some minor comments.

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: Rephrased this to mention getting new command-lists from UR queue

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: Done, i've also expanded this section quite a bit

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: Good point, gone with "Execution Order"

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: I've expanded this section to give more context and hopefully make it clearer, if you want to take another look

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: I've expanded this section to give more context and hopefully make it clearer, if you want to take another look

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md:134]: It will be able to enable other backends in future, hopefully CUDA this year, so when can add new subsections here when they are done.

I've added an extra sentence below to make that clear though

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: Agree that once the PR is merged we should update this to the direct link

**Review by julianmi — APPROVED** [review]

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:139]: I should document these drawbacks to our implementation highlighted by the maintainer
https://github.com/intel/llvm/pull/9992#discussion_r1246262690

File: sycl/doc/design/CommandGraph.md
Comment (mfrancepillois) [reply sycl/doc/design/CommandGraph.md:134]: Make sense.

File: sycl/doc/design/CommandGraph.md
Comment (mfrancepillois) [reply sycl/doc/design/CommandGraph.md]: Far much clearer. Thanks :+1:

**Review by mfrancepillois — APPROVED** [review]
LGTM

**Review by Bensuo — APPROVED** [review]

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
e.g. kernel execution, memory copy, host-task. In DPC++ an internal "finalization"
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
Future work will include exploring L0 API extensions to improve the mapping of 
UR command-buffer to L0 command-list.
```

**Comment by EwanC** [comment]
Now that the spec is merged, I've updated the spec link to point to the spec location rather than open PR. So I think this can be merged now and pushed back upstream

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
UR command-buffer, where a command-buffer object represents a series of operations
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
The SYCL Graph extension distinguishes between backends that support command-buffer extensions and
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
However during graph construction, inside `hander::finalize()` the
CG object is not submitted for execution as normal, but stored in the
graph as a new node instead.
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
workloads/hardware. In the future there might be options for controlling this
implementation detail by the user.
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
other backends are implemented.
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
In order to achieve the expected UR command-buffer enqueue semantics with Level Zero,
the adapter implementation adds extra commands to the Level Zero command-list
representing a UR command-buffer.
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
1. 3x the command-list resources are used, if there are many UR command-buffers in
   flight, this may exhaust L0 driver resources. A trivial graph requires 3 L0
   command-lists and if we implement partitioning a graph into multiple UR
   command-buffers, then each partition will contain 3 L0 command-lists.
```

**Review by reble — APPROVED** [review]
LGTM, only some minor changes to avoid using first person pronoun.

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [reply sycl/doc/design/CommandGraph.md]: I've removed the plural but didn't really like the way it was reading so I've modified the surrounding wording a little to read better.

File: sycl/doc/design/CommandGraph.md
Comment (Bensuo) [reply sycl/doc/design/CommandGraph.md]: Removed the `we` but reworked the wording slightly from the suggestion.

**Pull request merged**
